man/avg_chromakey.1
.TH AVG_CHROMAKEY "1" "March 2011" "libavg 1.5.4" "User Commands"
.SH NAME
avg_chromakey \- test application for libavg chromakey filter
.SH SYNOPSIS
.B avg_chromakey
\fI\-t DRIVER [OPTION]...\fR
.SH DESCRIPTION
Stream a live image from a camera and apply a chromakey filter.
.PP
.SH OPTIONS
.TP
\fB\-h\fR, \fB\-\-help\fR
Show a brief description of the options.
.TP
\fB\-t\fR, \fB\-\-driver\fR DRIVER
One of the camera subsystems supported by libavg (supported: firewire, video4linux, directshow).
.TP
\fB\-d\fR, \fB\-\-device\fR DEVICE
Camera device identifier (depending on the driver, a GUID or device path is expected).
.TP
\fB\-u\fR, \fB\-\-unit\fR UNIT
Unit number. Used for cameras or other capture devices which deliver several images (e.g. stereo cameras or TV capture cards).
.TP
\fB\-w\fR, \fB\-\-width\fR WIDTH
Camera image width.
.TP
\fB\-e\fR, \fB\-\-height\fR HEIGHT
Camera image height.
.TP
\fB\-p\fR, \fB\-\-pixelformat\fR PIXELFORMAT
Camera pixel format (one of: I8, I16, YUV411, YUV422, YUYV422, RGB, BGR, BAYER8).
.TP
\fB\-f\fR, \fB\-\-framerate\fR FRAMERATE
Frames per second.
.TP
\fB\-8\fR, \fB\-\-fw800\fR
Set firewire bus speed to s800 (if supported).
.SH AUTHOR
Ulrich von Zadow <coder@c-base.org>


NEWS
libavg 1.5.4 - February 28, 2011 (Linux only)

  svn revision 5644, branches/release1.5.x

  XInput 2.1 updates/fixes.

libavg 1.5.3 - February 20, 2011 (Linux only)

  svn revision 5599

  Major new feature:
    * Linux XInput 2.1 Multitouch support.

libavg 1.5.1 - November 14, 2010 (Mac only)

  svn revision 5275

  Fixed bug that caused fullscreen apps to crash on the mac (Bug # 162).

libavg 1.5.0 - November 6, 2010

  svn revision 5270

  Major New Features:
    * Nodes can be constructed directly from python, making createNode unnecessary.
    * Added support for rendering avg scenes to offscreen canvases.
    * Added GPU-based effects framework and Blur, Shadow, Chromakey effects.
    * Added multitouch driver model and support for Apple Magic Trackpads and TUIO 
      devices.
    * Support for flash videos with an alpha channel. Decoding is GPU-accellerated.

  Other Changes:
    * Added color controls (gamma, brightness, contrast) to all RasterNodes.
    * Node constructors have a parent parameter which makes a following addChild 
      redundant.
    * Added Player.setEventHook().
    * Added onscreen keyboard for touchscreens.
    * WordsNode.getNumLines(), .getCharIndexFromPos(), .getTextAsDisplayed().
    * DivNode.getEffectiveMediaDir().
    * Clearer profiling that works for multiple threads.
    * Added AVG_BREAK_ON_IMPORT and AVG_BREAK_ON_ASSERT env variables.
    * Added AVG_DUMP_TEST_FRAMES env variable.
    * Added Player.setWindowFrame().
    * Added VideoNode.queuelength.
    * Video stability & timing improvements.
    * DivNode crop default is now false.
    * Node class names are unified: All names end with 'Node': WordsNode, DivNode, etc.
    * Added ImageNode.compression for 16 bpp texture support.
    * Added avg_audioplayer.py.

    * Various compiler compatibility fixes: Compiles with gcc 4.5 and VC++ 2010 now.
    * Lots of bugfixes - see the bug tracker.

libavg 1.5.0.pre3 - November 4, 2010

  svn revision 5267

libavg 1.5.0.pre2 - November 3, 2010

  svn revision 5263

libavg 1.5.0.pre1 - October 31, 2010

  svn revision 5244 

libavg 1.0.0 - January 8, 2010

  svn revision 4489

  New Features:

    * hinting is now optional in words node.
    * Multitouch emulation module added to AVGMTApp.
    * masks now work with words nodes.
    * Added video.volume property and video.hasAudio() method.
    * Added AVG_LOG_CATEGORIES environment variable. Log configuration is now
      done using the environment and not by calling a method.
    * Added maskpos and masksize attribute so masks can be moved and rescaled.
    * Added Player.getKeyModifierState()
    * Added sse2 version of yuv->rgb color conversion.
    * Added avg_videoinfo.sh.
    * Added mesh node.
    * Added avg_showcamera --resetbus to reset the firewire bus after crashes.

  Other Improvements:
    * New anim framework is now stable.
    * Snow Leopard compatibility.
    * avg_showcamera has much better command line handling.
    * The test suite is faster and a lot more comprehensive.
    * hrefs and mediadirs are now unicode-aware, so filenames like "ö.png" 
      should work.
    * node.unlink(True) now deletes all event handlers in the node, making
      it a lot easier to avoid memory leaks. 
    * Lots of bug fixes:
      - Windows gamma is now reset correctly on program end.
      - Linux vsync now much more stable.
      - Setting node.pos.x now throws an error instead of being ignored. 
        Same for all other Point2D attributes.
      - video.getNumFrames() now returns the exact number of frames.
      - Behaviour of nodes before Player.play() and when they are not in the
        avg tree is now much more stable.
      - Fixed words node sensitive area for centered and right-aligned nodes.

libavg 1.0.0.pre3 - January 6, 2010

  svn revision 4475

libavg 1.0.0.pre2 - December 23, 2009

  svn revision 4398

libavg 1.0.0.pre1 - December 21, 2009

  svn revision 4387

libavg 0.9.0 - August 23, 2009

  svn revision 4100

  Major new features:

    * Support for vector node types: Line, PolyLine, Curve, Circle, Polygon,
      Rectangle. Vector nodes can be textured, with the closed vectors taking two
      textures: one for the outline, one for the fill. Vector drawing is GPU-based.
    * Plugin support. Plugins written in C++ can define complete new node types.
    * Powerful (but still experimental) new animation framework.
    * Unified & much more stable camera drivers. Consistent support for all 
      resolutions & framerates we could test. Support for more than one camera on all
      platforms. fw800 is supported.
    * Uniform support for aggregate types in node interfaces (node.pos, node.pivot, 
      ...)

  Other improvements:
    * gcc 4.4 and ffmpeg 0.5.0 compatibility.
    * Added letterspacing support to words node.
    * Threaded videos are now the default.
    * Improved video seek and loop performance.
    * New Node::getElementByPos() method that returns the node at that point on the
      screen.
    * Added avg.svnrevision and avg.svndate properties so the avg version can be 
      queried.
    * New avg_showfont.py utility to help font selection.
    * New avg_showcamera.py utility to help camera setup.
    * New avg.getMemoryUsage function that returns the resident set size.
    * New Words.addFontDir().
    * Added support for mipmapping node textures, including videos.
    * rasternode.maskhref to support static alpha masks for any videos, word nodes,
      etc.

  Syntax and semantic changes:
    * In the course of adding aggregate types for node attributes, several attributes
      have changed:
      - words.size is now called words.fontsize.
      - pivotx and pivoty have been replaced by pivot.
    * right-aligned and centered words nodes are positioned differently.
    * avgtrackerrc syntax has changed significantly; see src/avgtrackerrc.minimal.
    * Support for firewire cameras using libdc1394 ver. 1 has been phased out. Please 

src/utils/avg_chromakey.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# libavg - Media Playback Engine.
# Copyright (C) 2003-2008 Ulrich von Zadow
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
# Current versions can be found at www.libavg.de
#

import optparse
from libavg import avg, AVGApp
from libavg.ui import button
from libavg import parsecamargs

GUI_SIZE=(300, 200)

g_Player = avg.Player.get()

class Slider(avg.DivNode):
    def __init__(self, width, min, max, onChange, **kwargs):
        avg.DivNode.__init__(self, **kwargs)
        self.__onChange = onChange
        self.size = (width, 20)
        self.__min = min
        self.__max = max
        self.__val = min
        avg.LineNode(pos1=(7,14), pos2=(width-7,14), color="FFFFFF", strokewidth=2, 
                parent=self)
        self.__slider = avg.DivNode(pos=(0,0), size=(14,20), parent=self)
        avg.PolygonNode(pos=((1,0), (13,0), (7,18)), fillopacity=1, fillcolor="FFFFFF",
                color="808080", parent=self.__slider)
        self.__slider.setEventHandler(avg.CURSORDOWN, avg.MOUSE, 
                self.__onSliderDown)
        self.setEventHandler(avg.CURSORDOWN, avg.MOUSE, self.__onBarDown)
        self.__isDragging = False

    def getVal(self):
        return self.__val

    def setVal(self, val):
        self.__val = val
        self.__positionSlider()

    val = property(getVal, setVal)

    def __onSliderDown(self, event):
        self.__slider.setEventCapture()
        self.__slider.setEventHandler(avg.CURSORMOTION, avg.MOUSE, self.__onSliderMove)
        self.__slider.setEventHandler(avg.CURSORUP, avg.MOUSE, self.__onSliderUp)
        self.__isDragging = True
        self.__dragStartVal = self.__val

    def __onSliderMove(self, event):
        numPixelsMoved = float(event.pos.x-event.lastdownpos.x)
        self.__val = (self.__dragStartVal+numPixelsMoved/(self.size.x-14)
                *(self.__max-self.__min))
        self.__positionSlider()

    def __onSliderUp(self, event):
        self.__onSliderMove(event)
        self.__slider.releaseEventCapture()
        self.__slider.setEventHandler(avg.CURSORMOTION, avg.MOUSE, None)
        self.__slider.setEventHandler(avg.CURSORUP, avg.MOUSE, None)
        self.__isDragging = False

    def __onBarDown(self, event):
        if not(self.__isDragging):
            localPos = self.getRelPos(event.pos)
            ratio = (localPos.x-7)/(self.size.x-14)
            self.__val = self.__min+ratio*(self.__max-self.__min)
            print localPos, ", ", ratio, ", ", self.__val
            self.__positionSlider()

    def __positionSlider(self):
        if self.__val < self.__min:
            self.__val = self.__min
        elif self.__val > self.__max:
            self.__val = self.__max
        ratio = ((self.__val-self.__min)/(self.__max-self.__min))
        self.__slider.pos = (ratio*(self.size.x-14), 0)
        self.__onChange()


class FXSlider(avg.DivNode):
    def __init__(self, row, min, max, fxNode, fxAttrName, caption, isInt, **kwargs):
        avg.DivNode.__init__(self, **kwargs)
        avg.RectNode(pos=(0,8), size=(280,38), color="808080", strokewidth=2,
                parent=self)
        textBgRect = avg.RectNode(pos=(8,2), fillcolor="000000", fillopacity=1, 
                strokewidth=0, parent=self)
        caption = avg.WordsNode(pos=(10,0), text=caption, parent=self)
        textBgRect.size = caption.getMediaSize()+(4, 2)
        self.__words = avg.WordsNode(pos=(240,23), parent=self)
        self.__slider = Slider(220, min, max, self.__onSliderMove, pos=(15,20),
                parent=self)
        self.pos = (0, row*46)
        self.__fxNode = fxNode
        self.__fxAttrName = fxAttrName
        self.__caption = caption
        self.__isInt = isInt
        self.__slider.val = getattr(self.__fxNode, fxAttrName)

    def __onSliderMove(self):
        if self.__isInt:
            setattr(self.__fxNode, self.__fxAttrName, int(self.__slider.val))
            self.__words.text = "%i"%self.__slider.val
        else:
            setattr(self.__fxNode, self.__fxAttrName, self.__slider.val)
            self.__words.text = "%.2f"%self.__slider.val


class TextButton(button.Button):
    def __init__(self, text, **kwargs):
        size = kwargs["size"]
        upNode = avg.DivNode()
        avg.RectNode(size=size, fillcolor="FFFFFF", fillopacity=1, color="FFFFFF",
                parent=upNode)
        avg.WordsNode(pos=(4,3), text=text, color="000000", parent=upNode)
        downNode = avg.DivNode()
        avg.RectNode(size=size, fillcolor="000000", fillopacity=1, color="FFFFFF",
                parent=downNode)
        avg.WordsNode(pos=(4,3), text=text, color="FFFFFF", parent=downNode)
        kwargs["upNode"] = upNode
        kwargs["downNode"] = downNode
        button.Button.__init__(self, **kwargs)


def colorToString(colorTuple):
    s = "%02X%02X%02X"%colorTuple[:-1]
    return s

class Chromakey(AVGApp):
    def init(self):
        avg.RectNode(size=(options.width,options.height), fillcolor="FF0000", 
                fillopacity=1, strokewidth=0, parent=self._parentNode)
        self.__camNode = avg.CameraNode(driver=options.driver, device=options.device, 
                 unit=options.unit, fw800=options.fw800, 
                 capturewidth=options.width, captureheight=options.height, 
                 pixelformat=options.pixelFormat, framerate=options.framerate, 
                 width=options.width, height=options.height, parent=self._parentNode)
        self.__camNode.play()
        self.__filter = avg.ChromaKeyFXNode()
        self.__camNode.setEffect(self.__filter)
        self.__filter.color = "0000FF"
        self.__filter.htolerance = 0.05
        self.__filter.stolerance = 1.0 
        self.__filter.ltolerance = 1.0
        self.__filter.softness = 0.0

        self.__createGUI()

    def __createGUI(self):
        self.__guiDiv = avg.DivNode(pos=(options.width+10,10), parent=self._parentNode)

        self.__colorWords = avg.WordsNode(pos=(0,14), parent=self.__guiDiv)
        self.__colorWords.text = "Key Color: "+self.__filter.color
        self.__colorRect = avg.RectNode(pos=(200,12), size=(20, 20), 
                fillcolor=self.__filter.color, fillopacity=1, 
                color="FFFFFF", parent=self.__guiDiv)
        self.__camNode.setEventHandler(avg.CURSORDOWN, avg.MOUSE, 
                self.__onColorDown)

        FXSlider(1, 0.0, 1.0, self.__filter, "htolerance", "Hue Tolerance", 
                False, parent=self.__guiDiv)
        FXSlider(2, 0.0, 1.0, self.__filter, "stolerance", "Saturation Tolerance", 
                False, parent=self.__guiDiv)
        FXSlider(3, 0.0, 1.0, self.__filter, "ltolerance", "Lightness Tolerance", 
                False, parent=self.__guiDiv)
        FXSlider(4, 0.0, 1.0, self.__filter, "softness", "Softness", 
                False, parent=self.__guiDiv)
        FXSlider(5, 0, 8, self.__filter, "erosion", "Erosion", 
                True, parent=self.__guiDiv)
        FXSlider(6, 0.0, 1.0, self.__filter, "spillthreshold", "Spill Suppression", 
                False, parent=self.__guiDiv)

        TextButton(pos=(0,332), text="Whitebalance", size=(100,22), 
                clickHandler=self.__onWhitebalance, parent=self.__guiDiv)
        TextButton(pos=(110,332), text="Dump Config", size=(100,22), 
                clickHandler=self.__dumpConfig, parent=self.__guiDiv)

        FXSlider(9, 0, 500, self.__camNode, "shutter", "Shutter", 
                True, parent=self.__guiDiv)
        FXSlider(10, 128, 1023, self.__camNode, "gain", "Gain", 
                True, parent=self.__guiDiv)


    def __onColorDown(self, event):
        pos = self.__camNode.getRelPos(event.pos)
        bmp = self.__camNode.getBitmap()
        color = bmp.getPixel(pos)
        colorString = colorToString(color)
        self.__filter.color = colorString
        self.__colorWords.text = "Key Color: "+colorString
        self.__colorRect.fillcolor = colorString

    def __onWhitebalance(self, event):
        self.__camNode.setWhitebalance(
                self.__camNode.getWhitebalanceU(), self.__camNode.getWhitebalanceV())
        self.__camNode.doOneShotWhitebalance()

    def __dumpConfig(self, event):
        print "Camera:"
        print "  device=", self.__camNode.device
        print "  shutter=", self.__camNode.shutter
        print "  gain=", self.__camNode.gain
        print "  White Balance: (u=", self.__camNode.getWhitebalanceU(), ", v=", \
                self.__camNode.getWhitebalanceV()
        
        print "Chromakey:"
        print "  color=", self.__filter.color
        print "  htolerance=", self.__filter.htolerance
        print "  stolerance=", self.__filter.stolerance
        print "  ltolerance=", self.__filter.ltolerance
        print "  softness=", self.__filter.softness
        print "  erosion=", self.__filter.erosion
        print "  spillthreshold=", self.__filter.spillthreshold

parser = optparse.OptionParser()
parsecamargs.addOptions(parser)

(options, args) = parser.parse_args()
if options.driver is None:
    parser.print_help()
    print
    print "ERROR: at least '--driver' must be specified"
    exit()

resolution=(GUI_SIZE[0]+options.width, max(GUI_SIZE[1],options.height))

Chromakey.start(resolution=resolution)

